File: app/src/main/java/by/klnvch/link5dots/models/Room.java
/*
 * MIT License
 *
 * Copyright (c) 2023 klnvch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package by.klnvch.link5dots.models;

import com.google.firebase.database.Exclude;
import com.google.firebase.database.IgnoreExtraProperties;
import com.google.gson.Gson;

import java.io.Serializable;
import java.util.ArrayList;

import androidx.annotation.NonNull;
import androidx.room.ColumnInfo;
import androidx.room.Embedded;
import androidx.room.Entity;
import androidx.room.PrimaryKey;

@Entity(tableName = "rooms")
@IgnoreExtraProperties
public class Room implements Serializable {
    @PrimaryKey
    @NonNull
    private String key;
    @ColumnInfo(name = "state")
    private int state;
    @ColumnInfo(name = "timestamp")
    private long timestamp;
    @ColumnInfo(name = "dots")
    private ArrayList<Dot> dots;
    @Embedded(prefix = "user_1_")
    private User user1;
    @Embedded(prefix = "user_2_")
    private User user2;
    @ColumnInfo(name = "type")
    private int type;
    @ColumnInfo(name = "is_send")
    @Exclude
    private boolean isSend = false;
    @ColumnInfo(name = "is_test")
    private boolean isTest = false;

    public Room() {
    }

    @NonNull
    public static Room fromJson(@NonNull String json) {
        return new Gson().fromJson(json, Room.class);
    }

    @Override
    public boolean equals(Object obj) {
        return obj instanceof Room && ((Room) obj).key.equals(this.key);
    }

    @NonNull
    public String toJson() {
        return new Gson().toJson(this);
    }

    @NonNull
    public String getKey() {
        return key;
    }

    public void setKey(@NonNull String key) {
        this.key = key;
    }

    public int getState() {
        return state;
    }

    public void setState(int state) {
        this.state = state;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public ArrayList<Dot> getDots() {
        return dots;
    }

    public void setDots(ArrayList<Dot> dots) {
        this.dots = dots;
    }

    public User getUser1() {
        return user1;
    }

    public void setUser1(User user1) {
        this.user1 = user1;
    }

    public User getUser2() {
        return user2;
    }

    public void setUser2(User user2) {
        this.user2 = user2;
    }

    public int getType() {
        return type;
    }

    public void setType(int type) {
        this.type = type;
    }

    @Exclude
    public boolean isSend() {
        return isSend;
    }

    @Exclude
    public void setSend(boolean send) {
        isSend = send;
    }

    public boolean isTest() {
        return isTest;
    }

    public void setTest(boolean test) {
        isTest = test;
    }
}

File: shared/src/commonMain/kotlin/by/klnvch/link5dots/domain/models/Room.kt
/*
 * MIT License
 *
 * Copyright (c) 2025 klnvch
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package by.klnvch.link5dots.domain.models

import kotlinx.serialization.Serializable
import kotlin.js.ExperimentalJsExport
import kotlin.js.JsExport

@OptIn(ExperimentalJsExport::class)
@JsExport
interface IRoom {
    val key: String
    val time: Int
    val dots: List<Dot>
    val user1: IUser?
    val user2: IUser?

    fun move(dot: Dot): IRoom
    fun undo(): IRoom

    fun getDuration() = dots.lastOrNull()?.dt ?: 0
    fun getEndTime() = time + getDuration()
    fun getWinningLine() = dots.findWinningLine()
    fun isNotOver() = getWinningLine() == null
    fun isOver() = getWinningLine() != null
    fun isFree(p: Point) = dots.find { it.x == p.x && it.y == p.y } == null
}

fun IRoom.canMove(user: IUser?) = (if (dots.size % 2 == 0) user1 else user2) == user
fun IRoom?.canMove(n: Int) = if (this == null || this.isOver()) false else dots.size % 2 == n
fun IRoom?.isWon(n: Int) = if (this == null || this.isNotOver()) false else dots.size % 2 == n
fun IRoom.canUndo(user: IUser?) = (if (dots.size % 2 == 1) user1 else user2) == user
fun IRoom.isNotEmpty() = dots.isNotEmpty()
fun IRoom.lastPoint(): Point? = dots.lastOrNull()
fun IRoom.isNew() = dots.lastOrNull()?.dt == 0
val IRoom.size get() = dots.size
fun IRoom.isOwner(user: NetworkUser): Boolean = (user1 as? NetworkUser)?.id == user.id

@Serializable
data class Room(
    override val key: String,
    override val time: Int,
    override val dots: List<Dot>,
    override val user1: IUser?,
    override val user2: IUser?,
) : IRoom {
    override fun move(dot: Dot) = copy(dots = dots + dot)
    override fun undo() = copy(dots = dots.dropLast(1))
}

fun List<Point>.findWinningLine(): WinningLine? {
    if (size < 9) return null

    val lastDot = last()
    val lastIndex = lastIndex % 2

    val points = filterIndexed { i, _ -> i % 2 == lastIndex }

    // y = x + (py - px)
    // y = -x + (py + px)
    // y = py
    // x = px
    val line = points.findMaxLine { it.y == it.x + (lastDot.y - lastDot.x) }
        ?: points.findMaxLine { it.y == -it.x + (lastDot.y + lastDot.x) }
        ?: points.findMaxLine { it.y == lastDot.y }
        ?: points.map { it.invert() }.findMaxLine { it.y == lastDot.x }?.map { it.invert() }

    return if (line != null) WinningLineImpl(line) else null
}

private inline fun List<Point>.findMaxLine(predicate: (Point) -> Boolean): List<Point>? = this
    .filter { p -> predicate(p) }
    .sortedBy { it.x }
    .fold<Point, List<Point>>(emptyList()) { acc, next ->
        if (acc.isEmpty()) listOf(next)
        else if (acc.last().x + 1 == next.x) acc + next
        else if (acc.size >= 5) acc
        else listOf(next)
    }.takeIf { it.size >= 5 }
